### ISSUE TITLE
Distributed Array Shape Assumption

### ISSUE THREAD

**Issue by kejacobson** [OpenMDAO/OpenMDAO#2176]
### Summary of Issue

For distributed variables, OM [assumes all dimensions of the array except the first one are the same across procresses.](https://github.com/OpenMDAO/OpenMDAO/blob/master/openmdao/core/system.py#L1493). If this isn't true as in the example below, the error message says:

```
RuntimeError: <model> <class Group>: Global size of output 'adder.x' (900) does not agree with local shape (3, 200)
```
which isn't descriptive of the actual problem. Someone did comment that this assumption is made at the linked part of the code, but I don't see anything about this assumption in the docs. So, rather than a bug, it may be an undocumented expected behavior that doesn't give a clear error message?

### Issue Type

- [x] Bug

- [ ] Enhancement  
      POEM #___ <N/A or POEM number at https://github.com/OpenMDAO/POEMs>

### Description

The example below shows the error described in the summary. Switching n0 and n1 allows the code to run. For most arrays, it is probably safe to assume everything but the first dimension is the same, however there are some cases where that isn't true. The place I ran into this is structural mode shapes where I want the dimensions to be [num_modes ,num_points,n_dof], where the points are the variable that is different sizes across processors, in order to most efficiency access the data with an outer loop over the modes. So with either C or F ordering, I'd want to middle dimension to be distributed meaning I can't just switch to F order to align with OM's assumption about the shape. However, I could flatten everything to meet the assumption.

### Example

```python
import numpy as np
import openmdao.api as om
from mpi4py import MPI

class TwoDArrayAdder(om.ExplicitComponent):
    def initialize(self):
        self.options.declare('n0')
        self.options.declare('n1')

    def setup(self):
        n0 = self.options['n0']
        n1 = self.options['n1']
        self.add_input('x', shape=((n0, n1)), distributed=True)
        self.add_output('x_sum', shape=((n0, n1)), distributed=True)

    def compute(self, inputs, outputs):
        outputs['x_sum'] = np.sum(inputs['x'], axis=0)

n0 = 3
n1 = 100 if MPI.COMM_WORLD.rank ==0 else 200
prob = om.Problem()
ivc = prob.model.add_subsystem('ivc',om.IndepVarComp())
ivc.add_output('x', val = np.ones((n0, n1)), distributed=True)

prob.model.add_subsystem('adder',TwoDArrayAdder(n0=n0, n1=n1))
prob.model.connect('ivc.x','adder.x')

prob.setup()
prob.run_model()
```

### Environment

OpenMDAO 3.9.3.dev0

**Comment by DKilkenny** [comment]
The code assumes that all but the first dimension of the shapes local parts in a distributed variable must match. Internally openmdao uses the concept of a 'global_shape' in some places and that breaks unless the above assumption is true. We could potentially support this but would require a significant refactor of OpenMDAO. We will look into this further

**Comment by robfalck** [comment]
Resolved by #2192.
